### DISCUSSION TITLE
custom-elements.json metadata in Docusaurus

### DISCUSSION THREAD

**Discussion by malikfaizanhaider** [facebook/docusaurus#8722] [Q&A]
How I can rendercustom-elements.json metadata of web components in .mdx file.

### Example.

[component-metadata:sl-avatar]

**Comment by Advait251206** [comment]
Docusaurus does not natively understand `custom-elements.json` or tags like:

```
[component-metadata:sl-avatar]
```

That syntax isn’t part of MDX or Docusaurus—it’s usually something people implement themselves via a plugin or custom remark/rehype transform.

---

## Recommended approach

To render `custom-elements.json` (Web Components metadata), you typically need to:

### 1. Parse the JSON file

Example structure:
```json
{
  "modules": [
    {
      "declarations": [
        {
          "tagName": "sl-avatar",
          "properties": [],
          "attributes": []
        }
      ]
    }
  ]
}
```

You load it in a Docusaurus plugin or at build time.

---

### 2. Create a custom MDX component

Create something like:

```tsx
// src/components/ComponentMetadata.tsx

import React from 'react';
import metadata from '../../custom-elements.json';

export default function ComponentMetadata({ tag }) {
  const component = metadata.modules
    .flatMap(m => m.declarations)
    .find(c => c.tagName === tag);

  if (!component) return <p>Component not found: {tag}</p>;

  return (
    <div>
      <h3>{component.tagName}</h3>

      <h4>Attributes</h4>
      <ul>
        {component.attributes?.map(attr => (
          <li key={attr.name}>
            {attr.name} ({attr.type?.text})
          </li>
        ))}
      </ul>
    </div>
  );
}
```

---

### 3. Use it in `.mdx`

Instead of:
```
[component-metadata:sl-avatar]
```

Use:

```mdx
import ComponentMetadata from '@site/src/components/ComponentMetadata';

<ComponentMetadata tag="sl-avatar" />
```

---

## 4. (Optional) Make your custom syntax work

If you really want:

```
[component-metadata:sl-avatar]
```

You must create a **remark plugin**:

- Parse MDX AST
- Replace text node with `<ComponentMetadata tag="sl-avatar" />`

This is more advanced and requires:

- `remark`
- `unist-util-visit`
- Docusaurus plugin system

---

## Key point

Docusaurus only supports:

- MDX components
- React components
- Custom remark/rehype plugins

It does **not** support custom shortcode syntax out of the box.

---

## Best practice

✔ Use MDX components instead of custom brackets  
✔ Keep metadata parsing in React layer  
✔ Only use remark plugins if you need CMS-like syntax

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
